File: apps/api/src/migrate.ts
import fs from 'fs';
import path from 'path';
import pg from 'pg';

/**
 * The migration runner that ships inside the API image.
 *
 * `pnpm db:migrate` needs a source checkout, which a pull-based deployment
 * does not have: the VPS pulls images and never clones the repository. The
 * schema still has to be applied before the API will start, so the SQL travels
 * with the image and this entry point applies it:
 *
 *   docker compose -f deploy/docker-compose.yml run --rm migrate
 *
 * It stays a deliberate, separate step rather than something the API does on
 * boot. Migrations are the one part of a deployment that can destroy data, and
 * §18 of the deployment doc is explicit that they run under supervision, not
 * as a side effect of a container restarting.
 */
const databaseUrl = process.env.MIGRATE_DATABASE_URL || process.env.DATABASE_URL;
const sqlRoot = process.env.MIGRATIONS_DIR || path.resolve(process.cwd(), 'db');

async function applyDirectory(client: pg.Client, dir: string, label: string): Promise<number> {
  if (!fs.existsSync(dir)) {
    console.warn(`[migrate] ${label} directory not found at ${dir}, skipping.`);
    return 0;
  }
  const files = fs
    .readdirSync(dir)
    .filter((f) => f.endsWith('.sql'))
    .sort();
  for (const file of files) {
    console.log(`[migrate] ${label}: ${file}`);
    await client.query(fs.readFileSync(path.join(dir, file), 'utf-8'));
  }
  return files.length;
}

/**
 * Gives the RLS-bound application role its password, mirroring db/migrate.ts.
 * Migration 004 creates it NOLOGIN precisely so no credential lives in a
 * committed .sql file.
 */
async function grantAppRoleLogin(client: pg.Client): Promise<void> {
  const role = process.env.APP_DB_USER || 'factory_app';
  const password = process.env.APP_DB_PASSWORD;
  if (!password) {
    console.warn(
      `[migrate] APP_DB_PASSWORD is not set, so ${role} stays NOLOGIN. The API connects as that role; ` +
        'connecting as the bootstrap superuser instead would bypass row-level security.'
    );
    return;
  }
  const quoted = await client.query<{ ident: string }>('SELECT quote_ident($1) AS ident', [role]);
  const literal = await client.query<{ lit: string }>('SELECT quote_literal($1) AS lit', [password]);
  await client.query(`ALTER ROLE ${quoted.rows[0].ident} LOGIN PASSWORD ${literal.rows[0].lit}`);
  console.log(`[migrate] ${role} can now log in (NOSUPERUSER, NOBYPASSRLS).`);
}

async function main(): Promise<void> {
  if (!databaseUrl) {
    throw new Error('Set MIGRATE_DATABASE_URL (or DATABASE_URL) to the schema owner connection.');
  }

  const client = new pg.Client({ connectionString: databaseUrl });
  await client.connect();
  try {
    const migrations = await applyDirectory(client, path.join(sqlRoot, 'migrations'), 'migration');
    await grantAppRoleLogin(client);

    if (/^(1|true|yes)$/i.test(process.env.SEED_DEMO_DATA ?? '')) {
      await applyDirectory(client, path.join(sqlRoot, 'seeds'), 'seed');
    }

    console.log(`[migrate] ${migrations} migration file(s) applied.`);
  } finally {
    await client.end();
  }
}

main().catch((error) => {
  console.error('[migrate] failed:', error instanceof Error ? error.message : error);
  // A swallowed failure would let the API start against a half-applied schema.
  process.exit(1);
});

File: apps/api/internal/platform/migrate/migrate.go
// Package migrate is the migration runner that ships inside the API image.
//
// `pnpm db:migrate` needs a source checkout, which a pull-based deployment
// does not have: the VPS pulls images and never clones the repository. The
// schema still has to be applied before the API will start, so the SQL
// travels with the image and `fv migrate` applies it:
//
//	docker compose -f deploy/docker-compose.yml --profile migrate run --rm migrate
//
// It stays a deliberate, separate step rather than something the API does on
// boot. Migrations are the one part of a deployment that can destroy data,
// and §18 of the deployment doc is explicit that they run under supervision,
// not as a side effect of a container restarting.
package migrate

import (
	"context"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/jackc/pgx/v5"
)

// Options shape one run. Every field mirrors the environment variable the
// Node runner read, so a compose file written for it still works.
type Options struct {
	// DatabaseURL is the schema owner connection (MIGRATE_DATABASE_URL, else
	// DATABASE_URL).
	DatabaseURL string
	// Root holds migrations/ and seeds/ (MIGRATIONS_DIR, default ./db).
	Root string
	// AppRole and AppPassword give the RLS-bound application role its
	// password; migration 004 creates it NOLOGIN precisely so no credential
	// lives in a committed .sql file (APP_DB_USER, APP_DB_PASSWORD).
	AppRole     string
	AppPassword string
	// SeedDemoData applies seeds/ after the migrations (SEED_DEMO_DATA).
	SeedDemoData bool
	// Out receives the progress lines.
	Out io.Writer
}

// Run applies every migrations/*.sql in name order, grants the app role its
// login, and applies seeds/*.sql when the demo data is on. It returns the
// number of migration files applied; any failure stops the run, because a
// swallowed one would let the API start against a half-applied schema.
func Run(ctx context.Context, opts Options) (int, error) {
	if opts.DatabaseURL == "" {
		return 0, errors.New("set MIGRATE_DATABASE_URL (or DATABASE_URL) to the schema owner connection")
	}
	if opts.Root == "" {
		opts.Root = "db"
	}
	if opts.AppRole == "" {
		opts.AppRole = "factory_app"
	}
	if opts.Out == nil {
		opts.Out = io.Discard
	}
	conn, err := pgx.Connect(ctx, opts.DatabaseURL)
	if err != nil {
		return 0, err
	}
	defer conn.Close(context.Background())

	migrations, err := applyDirectory(ctx, conn, filepath.Join(opts.Root, "migrations"), "migration", opts.Out)
	if err != nil {
		return 0, err
	}
	if err := grantAppRoleLogin(ctx, conn, opts); err != nil {
		return 0, err
	}
	if opts.SeedDemoData {
		if _, err := applyDirectory(ctx, conn, filepath.Join(opts.Root, "seeds"), "seed", opts.Out); err != nil {
			return 0, err
		}
	}
	fmt.Fprintf(opts.Out, "[migrate] %d migration file(s) applied.\n", migrations)
	return migrations, nil
}

// applyDirectory runs each .sql file in name order. A file is one
// multi-statement script sent as a whole, as the Node runner sent it, so a
// migration's own transaction blocks keep their meaning.
func applyDirectory(ctx context.Context, conn *pgx.Conn, dir, label string, out io.Writer) (int, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			fmt.Fprintf(out, "[migrate] %s directory not found at %s, skipping.\n", label, dir)
			return 0, nil
		}
		return 0, err
	}
	var files []string
	for _, e := range entries {
		if !e.IsDir() && strings.HasSuffix(e.Name(), ".sql") {
			files = append(files, e.Name())
		}
	}
	sort.Strings(files)
	for _, file := range files {
		fmt.Fprintf(out, "[migrate] %s: %s\n", label, file)
		sql, err := os.ReadFile(filepath.Join(dir, file))
		if err != nil {
			return 0, err
		}
		if _, err := conn.Exec(ctx, string(sql)); err != nil {
			return 0, fmt.Errorf("%s %s: %w", label, file, err)
		}
	}
	return len(files), nil
}

func grantAppRoleLogin(ctx context.Context, conn *pgx.Conn, opts Options) error {
	if opts.AppPassword == "" {
		fmt.Fprintf(opts.Out, "[migrate] APP_DB_PASSWORD is not set, so %s stays NOLOGIN. The API connects as that role; connecting as the bootstrap superuser instead would bypass row-level security.\n", opts.AppRole)
		return nil
	}
	var ident, literal string
	if err := conn.QueryRow(ctx, `SELECT quote_ident($1), quote_literal($2)`, opts.AppRole, opts.AppPassword).Scan(&ident, &literal); err != nil {
		return err
	}
	if _, err := conn.Exec(ctx, `ALTER ROLE `+ident+` LOGIN PASSWORD `+literal); err != nil {
		return err
	}
	fmt.Fprintf(opts.Out, "[migrate] %s can now log in (NOSUPERUSER, NOBYPASSRLS).\n", opts.AppRole)
	return nil
}
